Handle .pb filenames and paths dynamically in PbFileFetcher

philipp-leitl opened the pull request:
The current version is limited to the usage of only the .pb files in the LTS and in addition it requires that the granularity is set to yearly, resulting in a filename which ends with the four digits of the year (_yyyy.pb). As our archiver appliance for the Belle II PXD is setup with a granularity of one file per month in the LTS (_yyyy_mm.pb), I needed a modification. This proposal now does not only cope for our specific case but should be able to find all required files for a PV across LTS, MTS and STS independently of the configured granularity of the archiver.

As far as I understand this patch completely keeps backwards compatibility.

Changes in tests/test_data.py and tests/test_js.py were only made to satisfy flake8/black rule checker.

I added several test functions to tests/test_pb.py which helped me to fix the logic of _get_pb_files(). But I'm pretty new to pytest and especially to mocking, so I guess there might be better solutions to avoid duplicated code.

Thank you very much in advance for your comments!

Commit:
pb files are now fetched from the storage independently from the configured granularity and across STS/MTS/LTS

Commit:
fixed self._root handling for the PbFileFetcher

Commit:
removed obsolte function _get_pb_file()

Commit:
polished comments and doc strings

Commit:
added test for new pb.PbFileFetcher._get_pb_files(), fixed formatting according to Black format checks

Commit:
fixed logic error in pb.PbFileFetcher._get_pb_files(), added one more test

Commit:
Merge branch 'main' into feature/fetch_pb_files_dynamically

Commit:
reworked tests for _get_pb_files and adjusted the function accordingly

Commit:
cleaned up comments and docs

Commit:
added test for PV with two colons

Commit:
included test for backwards compatibility of root path

aawdls commented on aa/pb.py:
```diff
@@ -295,13 +296,79 @@ class PbFileFetcher(fetcher.Fetcher):
+                pv_files.append(f)
+        return pv_files
+
+    def _create_datetime_for_pb_file(self, filepath):
```
This could be a static method or just a function at module level, because it doesn't use any knowledge of the `PbFileFetcher` object. This may make it easier to test.

aawdls commented on aa/pb.py:
```diff
@@ -295,13 +296,79 @@ class PbFileFetcher(fetcher.Fetcher):
+        # The filename can contain just the year or stepwise more information up to
+        # year, month, day, hour and minutes.
+        dates = (
+            re.search(r"\d{4}(_\d{2})?(_\d{2})?(_\d{2})?(_\d{2})?", filename)
```
```suggestion
            re.search(r"\d{4}(_\d{2}){0,4}", filename)
```

aawdls commented on aa/pb.py:
```diff
@@ -295,13 +296,79 @@ class PbFileFetcher(fetcher.Fetcher):
+            re.search(r"\d{4}(_\d{2})?(_\d{2})?(_\d{2})?(_\d{2})?", filename)
+            .group(0)
+            .split("_")
+        )
```
`re.search` might return none if the filename does not match your expected format. You should handle this.

philipp-leitl replied:
What do you prefer in this case? Raising an IOError or logging a warning? In the last case _get_pb_files() would simply return an empty list, maybe?

aawdls replied:
Yes logging a warning and returning an empty list sounds good to me.

aawdls commented on aa/pb.py:
```diff
@@ -295,13 +296,79 @@ class PbFileFetcher(fetcher.Fetcher):
+        # Make sure to give a least 3 arguments to datetime.datetime.
+        while len(dates) < 3:
+            dates.append(1)
+        return datetime.datetime(*dates, tzinfo=pytz.utc)
```
It would be good to guard against malformed input here too. For example, if one of the integers is out of range for a day / hour, you'll probably get an exception from this line.

aawdls commented on aa/pb.py:
```diff
@@ -295,13 +296,79 @@ class PbFileFetcher(fetcher.Fetcher):
+        containing data of the specified time window. The granularity setup of the
+        archiver for LTS/MTS/STS can deviate from the standard one."""
+        if end is None:
+            end = datetime.datetime.now().astimezone(pytz.utc)
```
You could use `utils.utc_now()`

aawdls commented on tests/test_pb.py:
```diff
@@ -142,22 +142,186 @@ def test_PbFetcher_get_raises_if_get_throws_HTTPError_not_404(
+        "root/STS/a/b/c/d_2001_02_03_04.pb",
+    ]
+
+    def side_effect(path_pv_dir):
```
If you defined this function at the module level (ie move it out of `test_PbFileFetcher_get_all_pb_files_of_pv(`), then you would not need to repeat it later.

aawdls commented on aa/pb.py:
```diff
@@ -295,13 +296,79 @@ class PbFileFetcher(fetcher.Fetcher):
+        pv_files_datetime = [self._create_datetime_for_pb_file(f) for f in pv_files]
+        # Find the files we need between start and end.
+        start_index, end_index = None, None
+        for i, file_date in enumerate(pv_files_datetime):
```
You may be able to make this a bit simpler if you do something like:
```
matching_paths = []
for file_path in pv_files:
    file_date = self._create_datetime_for_pb_file(file_path)
    if file_date >= start and file_date <= end:
        matching_paths.append(file_path)
```

philipp-leitl replied:
A perfect example of "thinking too complex" :sweat_smile: thanks for this simplification.
I just added the case where also start > file_date of the last file. Depending of the time delta and the set granularity there might be already data from the given time window in the file.

aawdls reviewed (changes requested):
Thanks for contributing this change! I've made some comments and suggestions in the code.

The code CI failure and the irregularities of black might be due to different versions being used or different config getting picked up. Let's look at that once we're happy with what the code is doing.

Commit:
recommended changes from PR:
- removed assumed LTS/MTS/STS root paths, instead accepting a list of root paths
- made _create_datetime_for_pb_file() static
- handle unexpected format of date information in file name
- globally defined DUMMY_FILES and side_effect_glob() in test_pb.py
- simplified method to find matching paths for given time window

Commit:
fix PbFileFetcher roots conversion into list

Commit:
fixed handling of wrong date format in filenames

philipp-leitl commented:
Sorry for the multiple pushes. I think the issues should be fixed now.

aawdls commented:
Just a quick note. Thanks for pushing the changes. Sorry for the delay in reviewing them. I'll try to do this in the next couple of days.

spreiter commented:
Hi @philipp-leitl and @aawdls ,

I was wondering about the status of this merge request? Is there something missing?

Greetings
